spec/controllers/users/omniauth_callbacks_controller_spec.rb

```diff
@@ -0,0 +1,170 @@
+describe Users::OmniauthCallbacksController do
+  
+  def mock_auth_hash
+    {
+      'provider':'google_oauth2',
+      'uid': '12345',
+      'info': {
+        'name':'Test user',
+        'email':'testgoogle@example.com'
+      }
+    }
+  end
+
+  def create_invalid_mock_hash(opts={})
+    new_hash = mock_auth_hash.merge(opts)
+    OmniAuth.config.add_mock(:google_oauth2, new_hash)
+    request.env['omniauth.auth'] = OmniAuth.config.mock_auth[:google_oauth2]
+  end
+
+  before(:each) do
+    
+    request.env["devise.mapping"] = Devise.mappings[:user]
+    OmniAuth.config.add_mock(:google_oauth2, mock_auth_hash())
+    request.env['omniauth.auth'] = OmniAuth.config.mock_auth[:google_oauth2]
+  
+    @user_with_uid_provider = {
+      email: "test123@example.com",
+      login: "test_google123",
+      unencrypted_password: "secret",
+      unencrypted_password_confirmation: "secret",
+      role: "contester",
+      name: "Test user",
+      city: "Айтос",
+      uid: "12345",
+      provider: "google_oauth2"
+    }
+
+    @user_with_email = {
+      email: "testgoogle@example.com",
+      login: "test_google123",
+      unencrypted_password: "secret",
+      unencrypted_password_confirmation: "secret",
+      role: "contester",
+      name: "Test user",
+      city: "Айтос"
+    }
+
+    User.delete_all
+  end
+
+  describe "#google_oauth2" do
+
+    context "user does not exist" do
+
+      it "should redirect to edit_user_path" do
+        post :google_oauth2
+        expect(response).to redirect_to(edit_user_path(User.last))
+      end
+
+      it "should successfully create a user" do
+        expect {
+          post :google_oauth2
+        }.to change{ User.count }.by(1)
+
+        user = User.last
```

Comment:
A more reliable way to handle this might be `User.find_by(email: mock[:info][:email])`